qubesadmin/tools/qvm_template.py

```diff
@@ -0,0 +1,409 @@
+parser.add_argument('--upgrades', action='store_true')
+
+# NOTE: Verifying RPMs this way is prone to TOCTOU. This is okay for local
+# files, but may create problems if multiple instances of `qvm-template` are
+# downloading the same file, so a lock is needed in that case.
+def verify_rpm(path, nogpgcheck=False, transaction_set=None):
+    if transaction_set is None:
+        transaction_set = rpm.TransactionSet()
+    with open(path, 'rb') as f:
+        try:
+            hdr = transaction_set.hdrFromFdno(f)
+            if hdr[rpm.RPMTAG_SIGSIZE] is None \
+                    and hdr[rpm.RPMTAG_SIGPGP] is None \
+                    and hdr[rpm.RPMTAG_SIGGPG] is None:
+                return nogpgcheck
+        except rpm.error as e:
+            if str(e) == 'public key not trusted' \
+                    or str(e) == 'public key not available':
+                return nogpgcheck
+            return False
+    return True
+
+def get_package_hdr(path, transaction_set=None):
+    if transaction_set is None:
+        transaction_set = rpm.TransactionSet()
+    with open(path, 'rb') as f:
+        hdr = transaction_set.hdrFromFdno(f)
+        return hdr
+
+def extract_rpm(name, path, target):
+    with open(path, 'rb') as in_file:
+        rpm2cpio = subprocess.Popen(['rpm2cpio', path], stdout=subprocess.PIPE)
+        # `-D` is GNUism
+        cpio = subprocess.Popen([
+                'cpio',
+                '-idm',
+                '-D',
+                target,
+                '.%s/%s/*' % (PATH_PREFIX, name)
+            ], stdin=rpm2cpio.stdout, stdout=subprocess.DEVNULL)
+        return rpm2cpio.wait() == 0 and cpio.wait() == 0
+
+def parse_config(path):
+    with open(path, 'r') as f:
+        return dict(line.rstrip('\n').split('=', 1) for line in f)
+
+def install(args, app):
+    # TODO: Lock, mentioned in the note above
+    transaction_set = rpm.TransactionSet()
+
+    rpm_list = []
+    for template in args.templates:
+        if template.endswith('.rpm'):
+            if not os.path.exists(template):
+                parser.error('RPM file \'%s\' not found.' % template)
+            rpm_list.append(template)
+
+    os.makedirs(args.cachedir, exist_ok=True)
+
+    dl_list = get_dl_list(args, app)
+    dl_list_copy = dl_list.copy()
+    # Verify that the templates are not yet installed
+    for name, ver in dl_list.items():
+        if name in app.domains:
+            print(('Template \'%s\' already installed, skipping...'
+                ' (You may want to use the {reinstall,upgrade,downgrade}'
+                ' operations.)') % name, file=sys.stderr)
+            del dl_list_copy[name]
+        else:
+            version_str = build_version_str(ver)
+            target_file = \
+                '%s%s-%s.rpm' % (PACKAGE_NAME_PREFIX, name, version_str)
+            rpm_list.append(os.path.join(args.cachedir, target_file))
+    dl_list = dl_list_copy
+
+    download(args, app, path_override=args.cachedir, dl_list=dl_list)
```

Comment:
To avoid accidental use of the file before verification, consider downloading to a differently-named file (like, append `.unverified` suffix or similar) and only rename after verification.